text/0024-schema-syntax.md

```diff
@@ -0,0 +1,396 @@
+# Cedar Schema Syntax
+
+## Related issues and PRs
+
+- Reference Issues: 
+- Implementation PR(s): 
+
+## Timeline
+
+- Start Date: 2023-07-24
+- Date Entered FCP:
+- Date Accepted:
+- Date Landed:
+
+## Summary
+
+This document proposes a custom syntax for Cedar schemas, as an alternative to its JSON-based syntax. 
+
+The syntax was developed with the following goals in mind:
+
+* Use the same concepts in both the JSON-based and custom syntax, allowing them to be naturally inter-converted
+* Reuse Cedar policy syntax as much as possible, to help with intuition
+* When no Cedar policy syntax analogue exists, leverage ideas from other programming languages that favor readability
+* Use as few syntactic concepts as possible; similar ideas in the schema should look similar syntactically
+
+## Basic example
+
+Here is the proposed syntax for the [TinyTodo schema](https://github.com/cedar-policy/cedar-examples/blob/main/tinytodo/tinytodo.cedarschema.json), whose auto-formatted JSON version is 160 lines long.
+```
+namespace "" {
+    entity Application;
+    entity User in [Team,Application] { name: String };
+    entity Team in [Team,Application];
+    entity List in [Application] {
+        owner: User,
+        name: String,
+        readers: Team,
+        editors: Team,
+        tasks: Set<{name: String, id: Long, state: String}>
+    };
+
+    action CreateList, GetLists 
+        appliesTo { principal: [User], resource: [Application] };
+
+    action GetList, UpdateList, DeleteList, CreateTask, UpdateTask, DeleteTask, EditShares 
+        appliesTo { principal: [User], resource:[List] };
+}
+```
+## Motivation
+
+Cedar schemas for non-toy applications are hard to read and write. The root cause of the difficulty is the use of JSON:
+
+* JSON has low information density. For three applications we’ve modeled — [TinyTodo](https://github.com/cedar-policy/cedar-examples/blob/main/tinytodo/tinytodo.cedarschema.json), [Document Cloud](https://github.com/cedar-policy/cedar-examples/blob/main/cedar-example-use-cases/document_cloud/schema.json), and [GitHub](https://github.com/cedar-policy/cedar-examples/blob/main/cedar-example-use-cases/github_example/schema.json) — schemas span 4-5 pages.
```

Comment:
Wouldn't yaml kind of solve these two issues?

Reply:
It would solve the problem with comments. I'm not sure about information density. And I'm not sure it's an improvement of a syntax that more closely matches Cedar itself. (Otherwise: Why didn't we use Yaml for Cedar?)